Jsonparser step2 bean

Kroisttt opened the pull request.

Kroisttt committed:
JsonArray 수정

타입 자체를 값으로 사용할 수 있게 수정

Kroisttt committed:
JsonObject 추가

Kroisttt committed:
JsonElement 지원 규격 추가

Kroisttt committed:
Tokenizer 수정

scanError() 추가

Kroisttt committed:
JsonType Object 추가

Kroisttt committed:
JsonElement 수정

Object 타입 지원, 변수명 수정

Kroisttt committed:
Parser 수정

JsonArray에서의 object 지원 추가, parse(object) 추가

Kroisttt committed:
JsonElement pair() 추가

Kroisttt committed:
JsonElement 수정

string 변수 doubleQuotion으로 수정

Kroisttt committed:
Parser 구조 수정

Kroisttt committed:
Parser 수정

Kroisttt committed:
JsonType 수정

불필요한 코드 제거

Kroisttt committed:
JsonElement colon 추가

";" 추가

Kroisttt committed:
OutputView 수정

Object 출력 지원

Kroisttt committed:
Parser 수정

토큰들을 문자열 형태로 저장했던 unfinishedToken 대신 배열 형태로 저장하게 수정

Kroisttt committed:
Parser 수정

Object parse 지원 추가

Kroisttt committed:
하드코딩 수정

godrm commented on JSONParser/JSONParser/Views/OutputView.swift:
```diff
@@ -12,6 +12,8 @@ struct OutputView {
             printArray(result)
+        } else if result is JsonObject {
+            printObject(result)
         }
```
if 구문으로 비교하지 않고 다형성으로 구현해보세요.

Kroisttt replied:
Countable 프로토콜을 추가해서 구현했습니다.

godrm commented on JSONParser/JSONParser/Parser.swift:
```diff
@@ -22,40 +24,106 @@ struct Parser {
+        
+        if pairResult != unfinishedTokens.last?.last {
+            if pairResult != nil {
+                return false
```
이 정도면 한꺼번헤 비교해도 될 것 같습니다.

Kroisttt replied:
수정했습니다.

godrm commented on JSONParser/JSONParser/Parser.swift:
```diff
@@ -22,40 +24,106 @@ struct Parser {
+        return try JsonObject(object: data)
+    }
+    
+    private static func isTokenFinish(_ unfinishedTokens: [String]) -> Bool {
```
함수 이름으로 어떤 동작을 처리하는지 명확하지 않습니다. 끝을 나타내는 토큰인지 확인하는건가요?

Kroisttt replied:
함수명을 수정했습니다.

godrm commented on JSONParser/JSONParser/Parser.swift:
```diff
@@ -64,10 +132,14 @@ struct Parser {
         
+        if token.first == JsonElement.startOfObject, token.last == JsonElement.endOfObject {
+            return try parse(object: tokens)
+        }
```
Object만 있고 Array가 없는 것은 내부에 중첩되는 게 Object만 있다는 가정인건가요?

Kroisttt replied:
이번 단계의 요구조건에서 내부에 Object만 중첩되기 때문에 Array는 구현하지 않았습니다.

godrm reviewed (commented):
구조적으로 잘 나눠져 있는데 일부 변수나 함수가 약간 어색합니다. 
다음 피드백을 확인해보세요.

Kroisttt committed:
Parser 수정

Kroisttt committed:
Countable 프로토콜 추가

Kroisttt committed:
Parser 수정

isTokenFinish() 대신 isTypeComplete(), unfinishedTokens 대신 incompleteTypeTokens로 수정
